### PR TITLE
Fast way to implement common interfaces

### PR DESCRIPTION
**Pull request by smotornyuk**
@amercader , hope you'll be interested in this pr - I just have some spare time and put what you suggested yesterday into code. 

TL;DR
It would be cool, to quickly implement `IActions`, `ITemplateHelpers`, etc, without using boilerplates in the form:
* add `plugin.implemens`
* add required method
* create some file with definitions
* import this file
* so on.

Instead, let's just quickly decorate plugin's class(explicitly, so there is no change in existing plugins and no magic - the same as Ian, I believe that magic ruins code);

Usage:
1. [Register everything](https://github.com/ckan/ckan/compare/master...DataShades:blanket-implementations?expand=1#diff-003078818bc68960df2fb3b068a82d0dR9).  Working on different extensions, I've noticed that people tend to use the same layout for introducing common entities: helpers reside in `helpers.py`, actions and auth functions in `logic.{action, auth}.py` or submodules, and for blueprints/cli we suggest to use `views.py` and `cli.py` respectively. If you are following such structure, just decorate your plugin with `toolkit.blanket_implementation()` and you'll get `ITemplateHelpers`, `IAuthFunctions`, `IActions`, `IBlueprint`, `IClick` as long as corresponding import paths available. These interfaces will use a dictionary(or list in case of blueprint/cli) of items, registered in `__all__` attribute of import path. i.e `ckanext.ext.views.__all__` defines registered blueprints.
2. [Cover single interface](https://github.com/ckan/ckan/compare/master...DataShades:blanket-implementations?expand=1#diff-003078818bc68960df2fb3b068a82d0dR14). If you want to do it in few iterations, just provide the first argument to decorator - it must be an item from `toolkit.Blanket` enumeration. Available options are `all`, `helper`, `auth`, `action`, `blueprint`, `cli`. `all` is used by default.
3. [Use custom import paths](https://github.com/ckan/ckan/compare/master...DataShades:blanket-implementations?expand=1#diff-003078818bc68960df2fb3b068a82d0dR19). Even if you don't follow the layout described above, just provide source module as the second argument to decorator and don't bother with the file structure.
4. [Add some dynamics with functions and lambdas](https://github.com/ckan/ckan/compare/master...DataShades:blanket-implementations?expand=1#diff-003078818bc68960df2fb3b068a82d0dR24). There are some cases, when one may want some actions available depending on some config value. In that case, put a function into the second slot of decorator and decide what to do in runtime.

Why "blanket"? Because it "covers" interface, covers implementation, covers the import path. I don't know - it's kind of cute and I'm pretty bad in naming.

### CODE REVIEW

**Comment by wardi** [comment]
Here's my suggested interface, instead of
```python
@tk.blanket_implementation(tk.Blanket.cli)
class ExampleBlanketCliPlugin(p.SingletonPlugin):
    pass
```
How about:
```python
@tk.blanket_cli
class ExampleBlanketCliPlugin(p.SingletonPlugin):
    pass
```
Then we just have `blanket_helper`, `blanket_auth`, `blanket_action`, `blanket_blueprint` defined the same way. This is explicit, concise, provides a clear pattern for future blankets and allows each blanket to have its own custom parameters

**Comment by smotornyuk** [comment]
I thought about something similar - after the discussion on Thursday having separate decorators has much more sense for me both in terms of extensibility and unambiguity(and I'm going to remove `all` option, whatever implementation we'll choose - it will only create more possibilities for incompatible changes every time blanket will change). it won't be difficult to add multiple decorators to class and it shouldn't look too cumbersome as long as decorators are simple.


 The only thing I'm concerning about when it comes to registering every blanket as toolkit item - won't there be an overwhelming number of toolkit members if we continue with adding everything there? So, instead of `tk.blanket_cli`, `tk.blanket.cli` form looks tidier from my point.

And some thoughts about the future evolution of this decorator. If there will be cases when the interface has few methods and we want to cover all of them, there should be some implementation aligned to single-method interfaces. For example, in the case of `IFacet`, it would be quite straightforward to create `facets.py` with three methods `[dataset_facets, group_facets, organization_facets]`, or pass as decorator argument function, that'll return required functions. But is there any sense to account such options, or let's from the very beginning pin blankets to only simplest single-method interfaces and other interfaces will be considered as requiring classical implementation?

Personally, even though it sounds like "cool, they allow us to write less code", I'm leaning restricted use-cases of blankets - initial idea was to provide a simple mechanism of reducing plain boilerplate. If this mechanism will be hard to get, will have some complex use-cases - we'll just end up with having two equal ways of doing the same thing, what I don't like at all.

**Comment by smotornyuk** [comment]
Updated as per @wardi 's suggestion. I've added documentation, and if anyone is too lazy to build it locally, here's how it looks:
blanket_docs

**Comment by wardi** [comment]
@amercader any comments on this?

File: ckan/plugins/blanket.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                subject = import_module(import_path)
+            except ImportError:
+                log.debug(
+                    u"Unable to import <%s> for "
+                    u"blanket implementation of %s for %s",
+                    import_path,
+                    self.interface().__name__,
+                    plugin.__name__,
+                )
+                return
```
Comment (wardi) [inline ckan/plugins/blanket.py:156]: i think I'd rather not hide this exception, it means the plugin is broken, right?

Comment (smotornyuk) [reply ckan/plugins/blanket.py]: Yes. I think I've used this catch initially block for `blanket.all` which should implement all available interfaces. And then we decided that `blanket.all` is too unpredictable, but I forgot about this block

**Comment by wardi** [comment]
@smotornyuk sorry I let this sit, can you fix up the conflict so we can merge. I like this feature

**Comment by smotornyuk** [comment]
No problem, I won't be able to use this feature in production till the official release anyway:)

**Review by amercader — COMMENTED** [review]
@smotornyuk Apologies for the long overdue review. This looks great. I added some minor comments.

And also rewrote the docs :upside_down_face: I think this gives a clearer idea for users on how to use them and removes internal implementation details, but feel free to pick only the bits you think are useful:

https://gist.github.com/amercader/857ee22e9ea1c913bee3133cae81b45f

File: ckan/plugins/blanket.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+    """
+
+    helper = enum.auto()
+    auth = enum.auto()
+    action = enum.auto()
+    blueprint = enum.auto()
+    cli = enum.auto()
+    validator = enum.auto()
+    _all = helper | auth | action | blueprint | cli | validator
```
[the remark this answers is not in the document]
Comment (smotornyuk) [reply ckan/plugins/blanket.py]: Yep, i'm done with testing so this line doesn't have any value anymore

File: ckan/plugins/blanket.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    def func(
+        self: p.SingletonPlugin, *args: Any, **kwargs: Any
+    ) -> Union[Dict[str, Any], List[Any]]:
+        if isinstance(subject, types.FunctionType):
+            return subject(*args, **kwargs)
+        elif isinstance(subject, types.ModuleType):
+            result = {
+                item: getattr(subject, item)
+                for item in getattr(subject, u"__all__", [])
+            }
```
[the remark this answers is not in the document]
Comment (smotornyuk) [reply ckan/plugins/blanket.py]: I've added a note about this point and updated logic a bit.

When `__all__` attribute is not defined:
  - all locally defined members whose name doesn't start with an underscore are exported
  
When `__all__` attribute is defined:
  - only items listed there are exported

File: ckan/plugins/blanket.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+   - ``ckanext.ext.logic.auth`` for ``IAuthFunctions``
+   - ``ckanext.ext.logic.action`` for ``IActions``
+   - ``ckanext.ext.logic.validators`` for ``IValidators``
+   - ``ckanext.ext.views`` for ``IBlueprint``
+   - ``ckanext.ext.cli`` for ``IClick``
+
+Available groups are:
+  - ``tk.blanket.helpers``: implemets ``ITemplateHelpers``
+  - ``tk.blanket.auth``: implemets ``IAuthFunctions``
+  - ``tk.blanket.action``: implemets ``IActions``
```
[the remark this answers is not in the document]
Comment (smotornyuk) [reply ckan/plugins/blanket.py]: Good point, updated

File: ckan/plugins/blanket.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amercader) [reply ckan/plugins/blanket.py]: Sounds good!

**Comment by amercader** [comment]
Thanks @smotornyuk for the changes!

blanket

**Comment by amercader** [comment]
Sorry @smotornyuk , not sure why but when merging this PEP8 tests failed in master. Can you have a look?

**Comment by smotornyuk** [comment]
Sure
